File: algorithms/0451.sort_characters_by_frequency/sort_characters_by_frequency.rs
impl Solution {
    pub fn frequency_sort(s: String) -> String {
        let mut chars: Vec<char> = s.chars().collect();
        let mut dict = [0u16; 128];
        let index = |ch| ch as usize - 30;
        for ch in s.chars() {
            dict[index(ch)] += 1;
        }
        chars.sort_by(|&a, &b| {
            if dict[index(b)] == dict[index(a)] {
                index(b).cmp(&index(a))
            } else {
                dict[index(b)].cmp(&dict[index(a)])
            }
        });
        // println!("{:?}", chars);
        chars.into_iter().collect()
    }
}

File: algorithms/0451.sort_characters_by_frequency/sort_characters_by_frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> alpha_freq;
        for (const char&c: s) {
            alpha_freq[c]++ ;
        }
        sort(s.begin(), s.end(), [&](const auto &a, const auto &b){
            return alpha_freq[a] == alpha_freq[b] ? a < b : alpha_freq[a] > alpha_freq[b];
        });
        return s;
    }
};